### PR TITLE
Flash Attention Conceptual Guide

### PR DESCRIPTION
**Pull request by merveenoyan**
PR for conceptual guide on flash attention. I will add more info unless I'm told otherwise.

### CODE REVIEW

File: docs/source/conceptual/flash_attention.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Flash Attention
+
+Scaling transformer architecture is heavily bottlenecked by the self-attention mechanism, which has quadratic time and memory complexity. Recent developments in accelerator hardware are mainly focused on enhancing compute capacities and not memory and transferring data between hardware. This results in attention operation having a bottleneck in memory, also called _memory-bound_. Flash Attention is an attention algorithm used to overcome this problem and scale transformer-based models more efficiently, enabling faster training and inference.
```
Comment (Narsil) [inline docs/source/conceptual/flash_attention.md:3]:
```suggestion
Scaling transformer architecture is heavily bottlenecked by the self-attention mechanism, which has quadratic time and memory complexity. Recent developments in accelerator hardware are mainly focused on enhancing compute capacities and not memory and transferring data between hardware. This results in attention operation having a bottleneck in memory, also called _memory-bound_. Flash Attention is an attention algorithm used to reduce this problem and scale transformer-based models more efficiently, enabling faster training and inference. 
```

File: docs/source/conceptual/flash_attention.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# Flash Attention
+
+Scaling transformer architecture is heavily bottlenecked by the self-attention mechanism, which has quadratic time and memory complexity. Recent developments in accelerator hardware are mainly focused on enhancing compute capacities and not memory and transferring data between hardware. This results in attention operation having a bottleneck in memory, also called _memory-bound_. Flash Attention is an attention algorithm used to overcome this problem and scale transformer-based models more efficiently, enabling faster training and inference. 
+In standard attention implementation, the cost of loading and writing keys, queries, and values from High Bandwidth Memory (HBM) is high. It loads key, query, value from HBM to GPU, performs a single step of the attention mechanism and writes it back to HBM, and repeats this for every singular step of the attention. Instead, Flash Attention loads keys, queries, and values once, fuses the operations of the attention mechanism and writes them back.
```
Comment (Narsil) [inline docs/source/conceptual/flash_attention.md:4]:
```suggestion
In standard attention implementation, the cost of loading and writing keys, queries, and values from High Bandwidth Memory (HBM) is high. It loads key, query, value from HBM to GPU on-chip SRAM., performs a single step of the attention mechanism and writes it back to HBM, and repeats this for every singular step of the attention. Instead, Flash Attention loads keys, queries, and values once, fuses the operations of the attention mechanism and writes them back. 
```

Both rams are on the GPU, but there's the global address RAM, and the chip local (L1 equivalent if you will) RAM.

Comment (Narsil) [reply docs/source/conceptual/flash_attention.md:4]: Maybe even link the paper directly from more complete info ? https://arxiv.org/pdf/2205.14135.pdf

Comment (merveenoyan) [reply docs/source/conceptual/flash_attention.md:4]: I made this previously to clarify, WDYT? image

Comment (Narsil) [reply docs/source/conceptual/flash_attention.md:4]: Just link the paper, it's complex enough I think :)

Comment (Narsil) [reply docs/source/conceptual/flash_attention.md:4]: The main thing is that it doesn't "solve" the quadratic issue, it just makes the operation more efficient on the hardware.

File: docs/source/conceptual/flash_attention.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# Flash Attention
+
+Scaling transformer architecture is heavily bottlenecked by the self-attention mechanism, which has quadratic time and memory complexity. Recent developments in accelerator hardware are mainly focused on enhancing compute capacities and not memory and transferring data between hardware. This results in attention operation having a bottleneck in memory, also called _memory-bound_. Flash Attention is an attention algorithm used to reduce this problem and scale transformer-based models more efficiently, enabling faster training and inference. 
+In standard attention implementation, the cost of loading and writing keys, queries, and values from High Bandwidth Memory (HBM) is high. It loads key, query, value from HBM to GPU on-chip SRAM, performs a single step of the attention mechanism and writes it back to HBM, and repeats this for every singular step of the attention. Instead, Flash Attention loads keys, queries, and values once, fuses the operations of the attention mechanism and writes them back.
```
Comment (osanseviero) [inline docs/source/conceptual/flash_attention.md:4]: Were you intending to add a new line here?

File: docs/source/conceptual/flash_attention.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+In standard attention implementation, the cost of loading and writing keys, queries, and values from High Bandwidth Memory (HBM) is high. It loads key, query, value from HBM to GPU on-chip SRAM, performs a single step of the attention mechanism and writes it back to HBM, and repeats this for every singular step of the attention. Instead, Flash Attention loads keys, queries, and values once, fuses the operations of the attention mechanism and writes them back. 
+It is implemented for models with custom kernels, you can check out the full list of models that support Flash Attention [here](https://github.com/huggingface/text-generation-inference/tree/main/server/text_generation_server/models). 
+You can learn more about Flash Attention by reading the paper in this [link](https://arxiv.org/abs/2205.14135).
+
```
Comment (osanseviero) [inline docs/source/conceptual/flash_attention.md:7]: Anything we could do to make https://moon-ci-docs.huggingface.co/docs/text-generation-inference/pr_892/en/conceptual/flash_attention more visual?

File: docs/source/conceptual/flash_attention.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# Flash Attention
+
+Scaling transformer architecture is heavily bottlenecked by the self-attention mechanism, which has quadratic time and memory complexity. Recent developments in accelerator hardware are mainly focused on enhancing compute capacities and not memory and transferring data between hardware. This results in attention operation having a bottleneck in memory, also called _memory-bound_. Flash Attention is an attention algorithm used to reduce this problem and scale transformer-based models more efficiently, enabling faster training and inference.
```
Comment (osanseviero) [inline docs/source/conceptual/flash_attention.md:3]:
```suggestion
Scaling the transformer architecture is heavily bottlenecked by the self-attention mechanism, which has quadratic time and memory complexity. Recent developments in accelerator hardware mainly focus on enhancing compute capacities and not memory and transferring data between hardware. This results in attention operation having a memory bottleneck, also called memory-bound. **Flash Attention** is an attention algorithm used to reduce this problem and scale transformer-based models more efficiently, enabling faster training and inference. 
```

File: docs/source/conceptual/flash_attention.md (hunk 3 above)
Comment (osanseviero) [inline docs/source/conceptual/flash_attention.md:4]:
```suggestion
In the standard attention implementation, the cost of loading and writing keys, queries, and values from High Bandwidth Memory (HBM) is high. It loads key, query, and value from HBM to GPU on-chip SRAM, performs a single step of the attention mechanism, writes it back to HBM, and repeats this for every single attention step. Instead, Flash Attention loads keys, queries, and values once, fuses the operations of the attention mechanism, and writes them back. 
```

File: docs/source/conceptual/flash_attention.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+In standard attention implementation, the cost of loading and writing keys, queries, and values from High Bandwidth Memory (HBM) is high. It loads key, query, value from HBM to GPU on-chip SRAM, performs a single step of the attention mechanism and writes it back to HBM, and repeats this for every singular step of the attention. Instead, Flash Attention loads keys, queries, and values once, fuses the operations of the attention mechanism and writes them back. 
+It is implemented for models with custom kernels, you can check out the full list of models that support Flash Attention [here](https://github.com/huggingface/text-generation-inference/tree/main/server/text_generation_server/models).
```
Comment (osanseviero) [inline docs/source/conceptual/flash_attention.md:5]:
```suggestion
It is implemented for models with custom kernels. You can check out the complete list of models that support Flash Attention [here](https://github.com/huggingface/text-generation-inference/tree/main/server/text_generation_server/models).
```

Comment (osanseviero) [inline docs/source/conceptual/flash_attention.md:5]: Do all these model support flash attention?

**Review by osanseviero — COMMENTED** [review]
I'm finding this conceptual guide a bit hard to read/parse. We throw many technical concepts (memory bound, flash attention, high bandwidth memory, SRAM, etc) in two paragraphs. I would suggest to simplify a bit the language or extend a bit more the guide to make sure readers get an ok understanding of flash attention after reading this.

File: docs/source/conceptual/flash_attention.md (hunk 4 above)
Comment (merveenoyan) [reply docs/source/conceptual/flash_attention.md:7]: I made this but it's a bit overwhelming, maybe I could change fonts & other things to make it prettier, WDYT? 
Memory

**Comment by merveenoyan** [comment]
@osanseviero @Reeray made it super pretty:

File: docs/source/conceptual/flash_attention.md (hunk 3 above)
Comment (merveenoyan) [reply docs/source/conceptual/flash_attention.md]: yep, I just added.

File: docs/source/conceptual/flash_attention.md (hunk 6 above)
Comment (merveenoyan) [reply docs/source/conceptual/flash_attention.md]: Added detail on prefix.

File: docs/source/conceptual/flash_attention.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+Scaling the transformer architecture is heavily bottlenecked by the self-attention mechanism, which has quadratic time and memory complexity. Recent developments in accelerator hardware mainly focus on enhancing compute capacities and not memory and transferring data between hardware. This results in attention operation having a memory bottleneck. **Flash Attention** is an attention algorithm used to reduce this problem and scale transformer-based models more efficiently, enabling faster training and inference. 
+
+Standard attention mechanism uses High Bandwidth Memory (HBM) to store, read and write keys, queries and values. HBM is large in memory, but slow in processing, meanwhile SRAM is smaller in memory, but faster in operations. In the standard attention implementation, the cost of loading and writing keys, queries, and values from HBM is high. It loads keys, queries, and values from HBM to GPU on-chip SRAM, performs a single step of the attention mechanism, writes it back to HBM, and repeats this for every single attention step. Instead, Flash Attention loads keys, queries, and values once, fuses the operations of the attention mechanism, and writes them back.
```
Comment (osanseviero) [inline docs/source/conceptual/flash_attention.md:5]: This is much much clearer to me. :fire: thanks a lot! :rock:

**Review by osanseviero — APPROVED** [review]
LGTM! Thanks a lot :fire: let's wait for @Narsil review as well

**Pull request merged**